core/src/Kokkos_MathematicalSpecialFunctions.hpp

```diff
@@ -1268,6 +1268,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
   return ch21;
 }
 
+template <auto BesselZero, auto BesselOne, class CmplxType, class IntType,
+          class... Args>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel(const IntType& a,
+                                            const CmplxType& z, Args... args) {
+  if (a == 0) return BesselZero(z, args...);
+  if (a == 1) return BesselOne(z, args...);
+  CmplxType bessel_im2 = BesselZero(z, args...);
+  CmplxType bessel_im1 = BesselOne(z, args...);
+
+  for (IntType cur_a = 2; cur_a <= a; ++cur_a) {
+    CmplxType tmp = bessel_im1;
+    bessel_im1    = CmplxType(2 * (cur_a - 1)) * bessel_im1 / z - bessel_im2;
+    bessel_im2    = tmp;
+  }
+
+  return bessel_im1;
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_j(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 25.0,
+                                              const IntType& bw_start   = 70) {
```

Comment:
In my opinion, these template arguments should only be exposed where they are directly used, i.e., in `cyl_bessel_j0` and `cyl_bessel_j1`.

Reply:
it is impossible to omit this parameters here, since templates and default arguments does not play nicely together, see:  https://godbolt.org/z/o4fG5d69f

Reply:
Right. I suggest specifying the necessary arguments in this function explicitly instead of exposing them as function parameters.

Reply:
I updated PR